### ISSUE TITLE
MirroredStrategy is not supported by ExponentialUpdateLossScaleManager and LossScaleOptimizer

### ISSUE THREAD

**Issue by etragas-fathom** [tensorflow/tensorflow#25080]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow):
Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):
Linux Ubuntu 16.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
No
- TensorFlow installed from (source or binary):
Binary
- TensorFlow version (use command below):
tf-nightly-gpu==1.13.0.dev20181022
- Python version:
3.6
- Bazel version (if compiling from source):
N/A
- GCC/Compiler version (if compiling from source):
N/A
- CUDA/cuDNN version:
Cuda 9.2
- GPU model and memory:
1x V100

----------
* TLDR: MirroredStrategy is not supported by [ExponentialUpdateLossScaleManager](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/mixed_precision/python/loss_scale_manager.py#L104) and [LossScaleOptimizer](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/mixed_precision/python/loss_scale_optimizer.py#L28)

* When attempting to use the LossScaleOptimizer with a MirroredStrategy, we get the following error: 

    * ```ValueError: Outputs of true_fn and false_fn must have the same type: int64, bool``` emanating from [this line](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/mixed_precision/python/loss_scale_optimizer.py#L158).
    We are able to get around this line by changing:
     ```is_overall_finite, true_apply_gradients_fn, gen_control_flow_ops.no_op)```
    To:
     ```is_overall_finite, true_apply_gradients_fn, lambda: tf.zeros([1], tf.int64))```

* Note, this error does not occur if MirroredStrategy is not enabled, and training proceeds as normal.
If we make the above change, then we can run the LossScaleOptimizer +  FixedLossScaleManager with MirroredStrategy enabled.

* However, if we attempt to use the ExponentialLossScaleManager  with MirroredStrategyEnabled instead of the FixedLossScaleManager we get the following:
```ValueError: You must specify an aggregation method to update a MirroredVariable in Tower Context".``` emanating from this line: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/mixed_precision/python/loss_scale_manager.py#L170

* This suggests that variables introduced and used by ExponentialLossScaleManager need to be aware of MirroredStrategy.  What exactly do we need to do to make sure ExponentialLossScaleManager variables can be handled by MirroredStrategy?  

* Our suspicion is that we need to handle variables from ExponentialLossScaleManager in _create_slots(), _prepare(), _apply_dense(), and _apply_sparse() as suggested by this part of the code in Optimizer https://github.com/tensorflow/tensorflow/blob/f9b9cf52eb36a5d5e8bbf96fc1f2b6f584ce7867/tensorflow/python/training/optimizer.py#L552
In addition, [this comment](https://github.com/tensorflow/tensorflow/issues/23986#issuecomment-444391385) (which has the same error as we do) suggested that reimplementing the optimizer is the right solution.

**Comment by etragas-fathom** [comment]
I'm not sure that is the right way to do things, but just wanted to flag.

**Comment by martinwicke** [comment]
contrib.mixed_precison will be replaced by a solution better integrated into Keras. These optimizers haven't been updated for DistributionStrategy and given contrib won't exist in 2.0 they won't. 

@reedwm @azaks2 FYI
